sphinx_panels/__init__.py

```diff
@@ -60,114 +184,70 @@ def run(self):
-                            footer_split,
-                        )
-                    )
-                start_line = i + 1
-                header_split = footer_split = None
-            if line.startswith("==="):
-                header_split = i - start_line
-            if line.startswith("..."):
-                footer_split = i - start_line
-            # TODO warn if header_split or footer_split not None
-        try:
-            panel_blocks.append(
-                (self.content[start_line:], start_line, header_split, footer_split)
-            )
-        except IndexError:
-            pass
+        panel_blocks = parse_panels(self.content, self.content_offset, default_classes)
 
+        # set the top-level containers
         parent = nodes.container(in_panel=True, classes=container_classes)
         rows = nodes.container(in_panel=True, classes=["row"])
         parent += rows
 
-        for (content, offset, header_split, footer_split) in panel_blocks:
-            classes = default_classes.copy()
-            for opt_offset, line in enumerate(content):
-                opt_match = RE_OPTIONS.match(line)
-                if not opt_match:
-                    break
-                if opt_match.group(2) == "+=":
-                    classes[opt_match.group(1)] = (
-                        classes[opt_match.group(1)] + opt_match.group(3).split()
-                    )
-                else:
-                    classes[opt_match.group(1)] = opt_match.group(3).split()
-
-            body_start = opt_offset
-            body_end = len(content)
-
-            if header_split is not None:
-                title_content = content[opt_offset:header_split]
-                title_offset = self.content_offset + offset + opt_offset
-                body_start = header_split + 1
-            else:
-                title_content = False
+        for data in panel_blocks:
 
-            if footer_split is not None:
-                footer_content = content[footer_split + 1 :]
-                footer_offset = self.content_offset + offset + footer_split
-                body_end = footer_split
-            else:
-                footer_content = False
-
-            body_content = content[body_start:body_end]
-            body_offset = self.content_offset + offset + body_start
+            classes = data["classes"]
 
             column = nodes.container(
                 in_panel=True, classes=["d-flex"] + classes["column"]
             )
             rows += column
             card = nodes.container(
-                in_panel=True, classes=["card", "w-100", "panel-card"] + classes["card"]
+                in_panel=True, classes=["card", "w-100"] + classes["card"]
             )
             column += card
 
-            if title_content:
-                title = nodes.container(
-                    in_panel=True, classes=["card-title"] + classes["title"]
+            if "header" in data:
+                header = nodes.container(
```

Comment:
one "gotcha" that we should make sure isn't an issue here is the word `container`. the pydata them realized that weird things were happening because `container` is a very common and specific bootstrap class, while Sphinx appends `container` to lots of things as well

Reply:
I specifically account for this in: https://github.com/executablebooks/sphinx-panels/blob/195e853a214df6234c107d5015b3213070bec42c/sphinx_panels/__init__.py#L302-L307